File: cohens_d.py
# cohens_d.py
"""
The cohens_d function will take two arrays and compute Cohen's d.

INPUT
x = ndarray
y = ndarray
DIM = the dimension the samples are along

OUTPUT
d = Cohen's d

Written by mvlombardo 05.09.2015
"""

def cohens_d(x, y, DIM='rows', SIGN=True):

    import numpy as np

    if DIM == 'rows':
        dim = 0
    elif DIM == 'columns':
        dim = 1

    # n-1 for x and y
    lx = x.shape[dim]-1
    ly = y.shape[dim]-1

    # mean difference
    if SIGN:
        md = np.mean(x, axis = dim, dtype = np.float64) - np.mean(y, axis = dim, dtype = np.float64)
    else:
        md = np.abs(np.mean(x, axis = dim, dtype = np.float64) - np.mean(y, axis = dim, dtype = np.float64))

    # pooled variance
    csd = (lx * np.var(x, axis = dim, dtype = np.float64)) + (ly * np.var(y, axis = dim, dtype = np.float64))
    csd = np.sqrt(csd/(lx + ly))

    # compute cohen's d
    d  = md/csd

    return(d)

File: cohens_d.m
function d = cohens_d(x,y,DIM)
% 
% Function will compute cohen's d effect size.
%
% INPUT
%	x = matrix or vector
%	y = matrix or vector
%	DIM = specify the dimension which samples are along
%
% Example usage
%
% x = [1:100]'
% y = [101:200]'
% d = cohens_d(x, y, 1)
%
% written by mvlombardo - 30.08.2015
%

% n-1 for x and y
lx = size(x,DIM)-1;
ly = size(y,DIM)-1;

% mean difference (numerator)
md = abs(nanmean(x,DIM) - nanmean(y,DIM));

% pooled variance (denominator)
csd = (lx .* nanvar(x,0,DIM)) + (ly .* nanvar(y,0,DIM));
csd = sqrt(csd./(lx + ly));

% Cohen's d
d = md./csd;

end % function d = cohens_d(x,y,DIM)